st3/lsp_utils/activity_indicator.py

```diff
@@ -0,0 +1,135 @@
+from abc import ABCMeta, abstractmethod
+from LSP.plugin.core.typing import Optional, Union
+from types import TracebackType
+from uuid import uuid4
+import sublime
+
+
+__all__ = ['ActivityIndicator']
+
+
+class StatusTarget(metaclass=ABCMeta):
+    @abstractmethod
+    def set(self, message: str) -> None:
+        ...
+
+    @abstractmethod
+    def clear(self) -> None:
+        ...
+
+
+class WindowTarget(StatusTarget):
+    def __init__(self, window: sublime.Window) -> None:
+        self.window = window
+
+    def set(self, message: str) -> None:
+        self.window.status_message(message)
+
+    def clear(self) -> None:
+        self.window.status_message("")
+
+
+class ViewTarget(StatusTarget):
+    def __init__(self, view: sublime.View, key: Optional[str] = None) -> None:
+        self.view = view
+        if key is None:
+            self.key = '_{!s}'.format(uuid4())
+        else:
+            self.key = key
+
+    def set(self, message: str) -> None:
+        self.view.set_status(self.key, message)
+
+    def clear(self) -> None:
+        self.view.erase_status(self.key)
+
+
+class ActivityIndicator:
+    """
+    An animated text-based indicator to show that some activity is in progress.
+
+    The `target` argument should be a :class:`sublime.View` or :class:`sublime.Window`.
+    The indicator will be shown in the status bar of that view or window.
+    If `label` is provided, then it will be shown next to the animation.
+
+    :class:`ActivityIndicator` can be used as a context manager.
+
+    .. versionadded:: 1.4
+    """
+    width = 10  # type: int
+    interval = 100  # type: float
+
+    _target = None  # type: Optional[StatusTarget]
+
+    def __init__(
+        self,
+        target: Union[StatusTarget, sublime.View, sublime.Window],
+        label: Optional[str] = None,
+    ) -> None:
+        self.label = label
```

Comment:
probably we can add a explicit method to change `label` rather than `xxx.label = 'new label'`?